### PR TITLE
[QoS] Add getNamespace to AtlasDBConfig

### PR DESCRIPTION
**Pull request by hsaraogi**
**Goals (and why)**: Add getNamespace so that the `QosClient` can use it. See #2650.

**Priority (whenever / two weeks / yesterday)**: Soon, small PR. Will cherry-pick and continue on the QoS client stuff, so not really blocking.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2661)

### CODE REVIEW

File: atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java [hunk 1]
```diff
@@ -313,17 +321,19 @@ private void checkNamespaceConfig() {
...
                                 + " DO NOT ATTEMPT TO FIX THIS YOURSELF.");
             }
+            return keyValueServiceNamespace;
         } else if (timelock().isPresent()) {
             // Special case - empty timelock and empty namespace/keyspace does not make sense
-            boolean timelockClientNonEmpty = !timelock().get().client().orElse("").isEmpty();
-            Preconditions.checkState(timelockClientNonEmpty,
+            Preconditions.checkState(timelock().get().client().isPresent(),
                     "For InMemoryKVS, the TimeLock client should not be empty");
+            return timelock().get().client().get();
```
Comment (jeremyk-91) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java]: nit: This last if block can probably be written `return timelock().get().client().orThrow(() -> new IllegalStateException("For InMemoryKVS..."));`. Up to you though

File: atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java [hunk 2]
```diff
@@ -35,6 +35,8 @@
 @Value.Immutable
 public abstract class AtlasDbConfig {
 
+    static final String UNSPECIFIED_NAMESPACE = "unspecifed";
```
[the remark this answers is not in the document]
Comment (jeremyk-91) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java]: Also should we add `@VisibleForTesting` above?

File: atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java [hunk 3]
```diff
@@ -313,17 +321,19 @@ private void checkNamespaceConfig() {
...
                         "AtlasDB refused to start, in order to avoid potential data corruption."
                                 + " Please contact AtlasDB support to remediate this. Specific steps are required;"
                                 + " DO NOT ATTEMPT TO FIX THIS YOURSELF.");
             }
+            return keyValueServiceNamespace;
         } else if (timelock().isPresent()) {
             // Special case - empty timelock and empty namespace/keyspace does not make sense
-            boolean timelockClientNonEmpty = !timelock().get().client().orElse("").isEmpty();
-            Preconditions.checkState(timelockClientNonEmpty,
+            Preconditions.checkState(timelock().get().client().isPresent(),
```
Comment (tboam) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java]: Think you've lost the case where the `client()` is present but is the empty String.  Was this intentional?

File: atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java [hunk 4]
```diff
@@ -313,17 +321,19 @@ private void checkNamespaceConfig() {
...
                     "For InMemoryKVS, the TimeLock client should not be empty");
+            return timelock().get().client().get();
         }
+        return UNSPECIFIED_NAMESPACE;
```
Comment (tboam) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java]: What's the situation where the namespace isn't specified at all?  Is it really valid?

File: atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java [hunk 5]
```diff
@@ -295,12 +302,13 @@ private void checkNamespaceConfig() {
...
                             "If present, the TimeLock client config should be the same as the"
                                     + " atlas root-level namespace config.")));
+            return namespace;
         } else if (!(keyValueService() instanceof InMemoryAtlasDbConfig)) {
             Preconditions.checkState(keyValueService().namespace().isPresent(),
                     "Either the atlas root-level namespace"
                             + " or the keyspace/dbName/sid config needs to be set.");
 
-            Optional<String> keyValueServiceNamespace = keyValueService().namespace();
+            String keyValueServiceNamespace = keyValueService().namespace().get();
```
Comment (tboam) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java]: You can minimise the diff by not extracting this until you return it.  But I don't mind

File: atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java [hunk 6]
```diff
@@ -283,6 +285,11 @@ private void checkLockAndTimestampBlocks() {
     }
 
     private void checkNamespaceConfig() {
+        getNamespaceString();
+    }
+
+    @Value.Derived
+    public String getNamespaceString() {
```
Comment (tboam) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java]: Is any of the config in here live reloadable?  Just wondering if there's a way we can only calculate this once rather doing this all the time.  Don't feel strongly about it though

**Comment by tboam** [comment]
Looks like legitimate test failures here.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java (hunk 3 above)
Comment (hsaraogi) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java]: An empty string here is treated as `Optional.empty()` by Jackson. Also, this is not the right place for this check, if we wanted to have constraint checks on the client string (like it does not contain an '@' or something) that should exist in `TimelockClientConfig`.

Comment (hsaraogi) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java]: Added a test for the empty string case.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java (hunk 4 above)
Comment (hsaraogi) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java]: This should only happen for inmemory config without timelock block.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java (hunk 6 above)
Comment (hsaraogi) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java]: Yes, this is not live-reloadable, feels like a good idea to not repeat the check multiple times.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java (hunk 5 above)
Comment (hsaraogi) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java]: But then we are doing the get multiple times :(

File: atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java (hunk 3 above)
Comment (hsaraogi) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java]: True, added it at the right place and a test.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (hsaraogi) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java]: Done.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java (hunk 1 above)
Comment (hsaraogi) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java]: Done.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java [hunk 7]
```diff
@@ -331,18 +327,26 @@ public String getNamespaceString() {
...
+                        .orElseThrow(() ->
+                                new IllegalStateException("For InMemoryKVS, the TimeLock client should not be empty"));
             }
             return UNSPECIFIED_NAMESPACE;
         }
     }
 
+    @Value.Derived
+    @JsonIgnore
+    public String getNamespaceString() {
```
Comment (tboam) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java]: this pattern for initialising fields isn't thread safe, I think we get away with it here but we should be careful

Comment (hsaraogi) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java]: Yes, agreed that this is an unsafe check-then-act pattern, but the reinitialization of the namespace should not be worrisome here.

**Pull request merged**